pkg/volume/util/operationexecutor/operation_generator_test.go

```diff
@@ -0,0 +1,202 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operationexecutor
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	fakeclient "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/record"
+	utiltesting "k8s.io/client-go/util/testing"
+	controllervolumetesting "k8s.io/kubernetes/pkg/controller/volume/attachdetach/testing"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/csi"
+	"k8s.io/kubernetes/pkg/volume/gcepd"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+	volumetesting "k8s.io/kubernetes/pkg/volume/testing"
+	volumetypes "k8s.io/kubernetes/pkg/volume/util/types"
+	"os"
+	"testing"
+	"time"
+
+	"fmt"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/csi-translation-lib/plugins"
+)
+
+// this method just tests the volume plugin name that's used in CompleteFunc, the same plugin is also used inside the
+// generated func so there is no need to test the plugin name that's used inside generated function
+func TestOperationGenerator_GenerateUnmapVolumeFunc_PluginName(t *testing.T) {
+	type testcase struct {
+		name                  string
+		isCsiMigrationEnabled bool
+		expectedPluginName    string
+	}
+
+	testcases := []testcase{
+		{
+			name:                  "csi migration disabled",
+			isCsiMigrationEnabled: false,
+			expectedPluginName:    plugins.GCEPDInTreePluginName,
+		},
+		{
+			name:                  "csi migration enabled",
+			isCsiMigrationEnabled: true,
+			expectedPluginName:    fmt.Sprintf("%s:%s", csi.CSIPluginName, plugins.GCEPDDriverName),
+		},
+	}
+
+	for _, tc := range testcases {
+		if tc.isCsiMigrationEnabled {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigration, true)()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigrationGCE, true)()
+		}
+		volumePluginMgr, plugin, tmpDir := initTestPlugins(t)
+		defer os.RemoveAll(tmpDir)
+
+		operationGenerator := getTestOperationGenerator(volumePluginMgr)
+
+		pdName := "pd-volume"
+		podName := "pod-1"
+		pod := getTestPodWithGCEPD(podName, pdName)
+		volumeToUnmount := getTestVolumeToUnmount(podName, pdName, pod)
+		expectedPluginName := tc.expectedPluginName
+
+		if tc.isCsiMigrationEnabled {
```

Comment:
I'm not sure what this entire block does. It seems like there is no result from this. the errors we're checking for in here could be other unit tests but I don't think need to be captured in the unmap test.

AFAIK `translateSpec` and `NewBlockVolumeMapper` would be implementations details of `GenerateUnmapVolumeFunc` which we're actually testing later anyway

Reply:
Actually I wrote this part to run the test, not to test the behavior of some functionality related with this part, because `GenerateUnmapVolumeFunc` call [`blockVolumePlugin.NewBlockVolumeUnmapper`](https://github.com/kubernetes/kubernetes/blob/3ccb63bb1b4d8a9e9a5e32bdd2750e44a6067ad4/pkg/volume/util/operationexecutor/operation_generator.go#L1198) and when the plugin is csi, csi plugin [looks a file](https://github.com/kubernetes/kubernetes/blob/3ccb63bb1b4d8a9e9a5e32bdd2750e44a6067ad4/pkg/volume/csi/csi_plugin.go#L699) that contains some information about the volume, and `GenerateUnmapVolumeFunc`fails if csi plugin can't find that file.

I call `plugin.NewBlockVolumeMapper` for csi enabled case to create that file.

Reply:
Thanks for the explanation, I get it now! Could you add a comment in the code explaining this. It's not clear from reading the code why `NewBlockVolumeMapper` is being called

Reply:
fixed, thanks!